OutOfSchool/OutOfSchool.WebApi.Tests/Services/NotificationServiceTests.cs

```diff
@@ -0,0 +1,270 @@
+                It.IsAny<Expression<Func<Notification, bool>>>(),
+                It.IsAny<string>()))
+            .Returns(Task.FromResult<IEnumerable<Notification>>(notifications));
+
+        var expectedGrouped = notifications
+            .Where(n => notificationsConfig.Grouped.Contains(n.Type.ToString()))
+            .GroupBy(n => new { n.Type, n.Action, n.GroupedData })
+            .Select(n => new NotificationGrouped
+            {
+                Type = n.Key.Type,
+                Action = n.Key.Action,
+                GroupedData = n.Key.GroupedData,
+                Amount = n.Count(),
+            })
+            .ToList();
+
+        var expectedGroupedByType = expectedGrouped
+            .GroupBy(n => n.Type)
+            .Select(n => new NotificationGroupedByType
+            {
+                Type = n.Key,
+                Amount = n.Sum(g => g.Amount),
+                GroupedByAdditionalData = n.ToList(),
+            })
+            .ToList();
+
+        var expectedNotifications = notifications
+            .Where(n => !notificationsConfig.Grouped.Contains(n.Type.ToString()))
+            .Select(notification => mapper.Map<NotificationDto>(notification))
+            .OrderByDescending(n => n.CreatedDateTime)
+            .ToList();
+
+        // Act
+        var result = await notificationService.GetAllUsersNotificationsGroupedAsync(userId).ConfigureAwait(false);
+
+        // Assert
+        notificationRepositoryMock.VerifyAll();
+        TestHelper.AssertTwoCollectionsEqualByValues(expectedNotifications, result.Notifications);
+        TestHelper.AssertTwoCollectionsEqualByValues(
+            expectedGroupedByType.Select(x => x.Amount),
+            result.NotificationsGroupedByType.Select(x => x.Amount));
+        Assert.AreEqual(
+            expectedGroupedByType.Select(x => x.GroupedByAdditionalData).Count(),
+            result.NotificationsGroupedByType.Select(x => x.GroupedByAdditionalData).Count());
+        for (int i = 0; i < expectedGroupedByType.Count; i++)
+        {
+            TestHelper.AssertTwoCollectionsEqualByValues(
+                expectedGroupedByType[i].GroupedByAdditionalData,
+                result.NotificationsGroupedByType.ElementAt(i).GroupedByAdditionalData);
+        }
+    }
+    #endregion
+
+    #region GetAllUsersNotificationsByFilterAsync
+    [Test]
+    public async Task GetAllUsersNotificationsByFilterAsync_WithValidUserIdAndNotificationType_ShouldReturnFilteredNotifications()
+    {
+        // Arrange
+        var notificationType = NotificationType.Parent;
+        notificationRepositoryMock
+            .Setup(x => x.GetByFilter(
+                It.IsAny<Expression<Func<Notification, bool>>>(),
+                It.IsAny<string>()))
+            .Returns(Task.FromResult<IEnumerable<Notification>>(notifications));
+        var expected = notifications
+            .Select(n => mapper.Map<NotificationDto>(n))
+            .Where(n => n.Type == notificationType)
```

Comment:
(Performance tiny issue) Please filter before mapping (use `.Where()` before `.Select()`), to avoid excessive work.

Reply:
Fixed. Thank you!